dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -0,0 +1,281 @@
+                result.Log.LogError(
+                    result.LastException,
+                    "Something went wrong in plan step {0}.{1}:'{2}'", this.SkillName, this.Name, context.LastErrorDescription);
+                return result;
+            }
+
+            context.Variables.Update(result.Result.ToString());
+        }
+        else
+        {
+            // loop through steps and execute until completion
+            while (this.HasNextStep)
+            {
+                var functionContext = context;
+                // Loop through State and add anything missing to functionContext
+
+                foreach (var item in this.State)
+                {
+                    if (!functionContext.Variables.ContainsKey(item.Key))
+                    {
+                        functionContext.Variables.Set(item.Key, item.Value);
+                    }
+                }
+
+                await this.InvokeNextStepAsync(functionContext);
+
+                context.Variables.Update(this.State.ToString());
+            }
+        }
+
+        return context;
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction SetDefaultSkillCollection(IReadOnlySkillCollection skills)
+    {
+        return this.Function is null
+            ? throw new NotImplementedException()
+            : this.Function.SetDefaultSkillCollection(skills);
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction SetAIService(Func<ITextCompletion> serviceFactory)
+    {
+        return this.Function is null
+            ? throw new NotImplementedException()
+            : this.Function.SetAIService(serviceFactory);
+    }
+
+    /// <inheritdoc/>
+    public ISKFunction SetAIConfiguration(CompleteRequestSettings settings)
+    {
+        return this.Function is null
+            ? throw new NotImplementedException()
+            : this.Function.SetAIConfiguration(settings);
+    }
+
+    #endregion ISKFunction implementation
+
+    /// <summary>
+    /// Invoke the next step of the plan
+    /// </summary>
+    /// <param name="context">Context to use</param>
+    /// <returns>The updated plan</returns>
+    /// <exception cref="KernelException">If an error occurs while running the plan</exception>
+    public async Task<Plan> InvokeNextStepAsync(SKContext context)
+    {
+        if (this.HasNextStep)
+        {
+            var step = this.Steps[this.NextStep];
+
+            context = await step.InvokeAsync(context);
+
+            if (context.ErrorOccurred)
+            {
+                throw new KernelException(KernelException.ErrorCodes.FunctionInvokeError,
+                    $"Error occurred while running plan step: {context.LastErrorDescription}", context.LastException);
+            }
+
+            this.NextStep++;
+            this.State.Update(context.Result.Trim());
+        }
+
+        return this;
+    }
+
+    public void SetFunction(ISKFunction function)
+    {
+        this.Function = function;
+        this.Name = function.Name;
```

Comment:
Do we need to have a specific property in the Plan, can't the property check and return the function directly, so we don't have to manage the Plan properties integrity?

Reply:
We need the properties as Plan implements ISKFunction. I could however create private fields for them and have the public interface return the Function property if Function is defined otherwise the private field. Thoughts?